src/keri/mod.rs

```diff
@@ -1,71 +1,223 @@
+    ///
+    pub fn incept_with_extra_keys(&mut self, extra_keys: impl IntoIterator<Item = (Basic, PublicKey)>)
+        -> Result<SignedEventMessage, Error> {
+            let mut keys: Vec<BasicPrefix> = extra_keys
+                .into_iter()
+                .map(|(key_type, key)| key_type.derive(key)).collect();
+            // Signing key must be first
+            keys.insert(0, Basic::Ed25519.derive(self.key_manager.borrow().public_key()));
+            let icp = EventMsgBuilder::new(EventType::Inception)?
+                .with_prefix(self.prefix.clone())
+                .with_keys(keys)
+                .with_next_keys(vec!(Basic::Ed25519.derive(self.key_manager.borrow().next_public_key())))
+                .build()?;
+
+            let signed = icp.sign(PayloadType::MA, vec!(
+                AttachedSignaturePrefix::new(
+                    SelfSigning::Ed25519Sha512,
+                    self.key_manager.borrow().sign(&icp.serialize()?)?,
+                    0
+                )
+            ));
+            let serialized = signed.serialize()?;
+            self.processor.process(signed_message(&serialized).unwrap().1)?;
+            self.prefix = icp.event.prefix;
+
+            Ok(signed)
+    }
+
+    /// Interacts with peer identifier via generation of a `Seal`
+    /// Seal gets added to our KEL db and returned back as `SignedEventMessage`
+    ///
+    pub fn interact(&self, peer: IdentifierPrefix)
+        -> Result<SignedEventMessage, Error> {
+        let next_sn = match self.processor.db.get_kel_finalized_events(&self.prefix) {
+            Some(mut events) => 
+                match events.next_back() {
+                    Some(db_event) => db_event.signed_event_message.event_message.event.sn,
+                    None => return Err(Error::InvalidIdentifierStat)
+                },
+            None => return Err(Error::InvalidIdentifierStat)
+        };
+        let (pref, seal) = match peer {
+            IdentifierPrefix::SelfAddressing(pref) =>
+                (pref.clone(), Seal::Digest(DigestSeal{dig: pref})),
+            _ => return Err(Error::SemanticError("Can interact with SelfAdressing prefixes only".into()))
+        };
+        let event = EventMessage::new(
+            Event::new(
+                self.prefix.clone(),
+                next_sn,
+                EventData::Ixn(InteractionEvent::new(pref, vec!(seal)))
+            ),
+            SerializationFormats::JSON)?;
+        let serialized = event.serialize()?;
+        let signature = self.key_manager.borrow().sign(&serialized)?;
+        let asp = AttachedSignaturePrefix::new(
+            SelfSigning::ECDSAsecp256k1Sha256,
+            signature,
+            0 // TODO: what is this?
+        );
+        let signed = SignedEventMessage::new(&event, PayloadType::OC, vec!(asp));
+        self.processor.db.add_kel_finalized_event(signed.clone(), &self.prefix)?;
```

Comment:
Why not use `process` of `process_actual_event` here? It will check if event is ok before adding it to kel.

Reply:
we've JUST created and signed it ourselves. it is not coming from "outside". what's the point of spending resources on re-checking our own product?

Reply:
It seems that we put to database event with wrong sn here. `process` will return error, that's why I proposed it. But you are right, we don't need re-checking.

Reply:
If processor did not manage to process same `ixn` event `sn` - mb we should tweak processor to be able to work properly? depends on above discussion outcome I guess.